### PR TITLE
Allow dllexport macro to be specified for grpc server side classes

### PR DESCRIPTION
**Pull request by IIFE**
@yashykt @markdroth @veblush @jtattermusch 

Similar to how `protoc` allows a dllexport macro to be specified for protobuf messages generated classes, this change allows generated server side grpc classes to be exported from a DLL if the user compiles the generated grpcs into a DLL (this is **not** exporting gRPC++ library classes, it's the generated code).

Should resolve ##937 as well.

The change does not affect existing usage as the new option must be specified for the macro to be inserted before the class name.

This is my first gRPC PR, so I'm not sure what the requirement for the `Mergable` check is.

Thanks.

### CODE REVIEW

**Comment by veblush** [comment]
Thanks for the PR! Can you run `tools/distrib/sanitize.sh` for reformatting your change and sign EasyCLA?

**Comment by veblush** [comment]
Don't the stub and method classes need the same thing?

**Comment by IIFE** [comment]
@veblush Thanks.

I'm having some difficulties running the `sanitize.sh` script as I'm on a Windows environment. Tried running it through git bash, but it throws various errors from the Python scripts.

No the stub doesn't need it because there is an interface available for it (`StubInterface`).

There is no harm in adding it to the `Stub` class itself, but didn't feel it added anything given the existence of `StubInterface`.

Thanks.

**Comment by IIFE** [comment]
@veblush 

I've managed to run the sanitize.sh script. However, it's modifying too many files, which I haven't touched as part of this PR. What should I do in this case?

P.S - I had to disable the docker option because I don't have docker installed (I'm using a Linux VM).

Thanks.

**Comment by veblush** [comment]
@IIFE Those scripts are not friendly to run on Windows. You might be able to fix your PR based on the result of sanity test as well. For this, can you sign up EasyCLA?

**Comment by jtattermusch** [comment]
Adding new protoc plugin compiler options is not to be taken lightly (and they need to be maintained and documented).
Isn't there some documentation that needs to be updated as well.
Can you link to the corresponding protoc compiler option in the protobuf codebase? (if we want to introduce the "same" option that already exists in protoc we need to double check the new option in the protoc plugin behaves the some / consistenly).

**Comment by IIFE** [comment]
@jtattermusch See [here ](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/compiler/cpp/cpp_generator.cc)for protobuf compiler option.

I couldn't find docs for the plugin generator. If you point me to it I can update it.

Same for tests. Can't find test files for the plugin.

This change is **not** generating extra constructors. It's simply marking the generated classes to be exported from a DLL. It needs to mark the copy ctor and assignment op as deleted because exporting makes the compiler generate default impls for those two functions. Those functions don't exist when the class is not exported, so deleting them when it's exported is not introducing different behaviour.

If you point me in the right direction, happy to add tests too.

**Comment by jtattermusch** [comment]
Based on a related discussion here: https://github.com/grpc/grpc/pull/25106#issuecomment-758713148 I'm not sure it's worth the effort to support dynamic library builds on windows (or we can make them "use at your own risk" level of support).

**Comment by IIFE** [comment]
@jtattermusch I'm not building gRPC libraries as a DLL., but rather building the generated code into a DLL. I am not concerned about the ABI and memory allocation issues you referenced because I build all my code using the same compiler version, so I can control those issues.

Given that protobuf supports the dllexport option (which solves the problem for the generated protobuf message classes), I think it's important that the grpc plugin supports the same option for the generated services classes. It's unfeasible for me to use static libs with what I'm working one, so being able to export the generated services from a DLL is very important.

As a side note, I think it's worthwhile for gRPC team to look into allowing gRPC (C++ layer) to be built as a DLL, and let users choose whether they want to use that option or not. The issues you referenced are not a problem if the same compiler version is used, which is something that users can ensure themselves. The referenced issues are even less of an issue for Windows because from Visual Studio 2015 the VS compiler has been backward compatible.

**Comment by jtattermusch** [comment]
Also see https://github.com/grpc/grpc/pull/25300

**Comment by IIFE** [comment]
Hi @veblush @jtattermusch - Is it possible to progress this PR please? Is there anything specific I need to do? I should stress again this change is not about dynamic build for the grpc library, but rather allowing the generated grpc doe to be exported from a DLL of the user's choice.

Thanks.

**Comment by jtattermusch** [comment]
@IIFE   
I understand that this option might come handy in your use case, but here's why I'm hesitant to include it:
- without tests that try to generate code with this option and then make sure it actually works on windows when protobufs are included from a DLL, there is no way of knowing if your patch works. At the same time, setting up such tests is non trivial amount of work and such tests require maintenance (and we don't have the bandwidth to do maintain them)
- since we're not really inclined to building grpc as a DLL (there are many problems), I guess the same holds true for importing protobufs from a DLL. There are simply too many scenarios that would need to be testing and how where things can break and realistically we can't support them all.
- It wouldn't make sense to include something that we'd have trouble supporting in the future (if there's a problem, who's going to fix it?) and IMHO including something that's potentially broken and it's hard to satisfactorily does more harm than good.

Hope that makes sense. I'm going to close this PR. As a workaround, you can simply apply the patch you came up with and use that in your codegen - for C++ this works pretty well since there is no codegen integration in the build system (as opposed to e.g. C# or python)

**Comment by IIFE** [comment]
I'm afraid I don't believe those points make sense.

Firstly, Protobuf messages already have an option to export their generated classes from a DLL. This is not a new thing.

Secondly, your first point can be applied to any change applicable to the grpc plugin. So are you saying you won't make any future changes to the plugin because it's hard to write tests for it?

Applying a patch myself is not reliable because the codegen can change at any point, causing me maintenance issues.

And finally, the changes proposed here are opt-in. Existing consumers won't be affected by it.

Thanks.

**Comment by jtattermusch** [comment]
> I'm afraid I don't believe those points make sense.
> 
> Firstly, Protobuf messages already have an option to export their generated classes from a DLL. This is not a new thing.
> 
> Secondly, your first point can be applied to any change applicable to the grpc plugin. So are you saying you won't make any future changes to the plugin because it's hard to write tests for it?

For a high priority feature we would take the time to write / setup tests. In this case it's a combination of a low priority feature and difficult to add tests (since we'd need to setup windows based tests that sufficiently exercise the "proto from DLL" scenario and that's quite different from some of the tests that we already have). At the same time, you haven't provided any automated tests that we could integrate yourself and we'd rather not accept a feature on "works for me" basis.

> 
> Applying a patch myself is not reliable because the codegen can change at any point, causing me maintenance issues.

I get that shifting the maintenance burden to us would be convenient for you, but I already explained why accepting the maintenance burden just doesn't make sense (the feature doesn't give us much and it's expensive to maintain and test - it's a simple cost/benefit analysis).

> 
> And finally, the changes proposed here are opt-in. Existing consumers won't be affected by it.

We don't want to have arbitrary opt-in features lying around in our codebase without testing them (such features tend to break a lot). This feature being opt-in and windows-specific is precisely what's making it error prone and hard to test. If this was a codepath that gets sufficiently exercised in almost every build and all users would use it, it would be a different story.

> 
> Thanks.

**Comment by IIFE** [comment]
We'll have to agree to disagree on the cost/benefit part.

The problem is you're essentially alienating the majority of your Windows users that work with C++. Most large-scale C++ applications on Windows rely on DLLs. The fact there are many issues (too many to link here) on this repo all related to problems either with using GRPC C++ as a DLL, or using it in a DLL environment, is a good hint there is something to be addressed. As a matter of fact, there appears to already be a similar PR to this one at #17591.

There is a clear need and benefit to consumers.

Thanks.

**Comment by jtattermusch** [comment]
If you look at the diff between https://github.com/grpc/grpc/pull/17591 and this PR, you'll see that the diff is radically different even though they are trying to achieve the same thing. That means one of them is broken and it might be broken in a subtle way. How do I know which one of them is broken and which one works? (or perhaps both are broken and a different diff would be needed?). This only stresses that fact that we'd need a good set of tests for this - neither of the PRs offer any tests nor any in-depth explanation why the __dllexport is needed to be added exactly in the places where it's being suggested (and/or has a comparison of what the protoc plugin does). Both PRs act if this is just a minor change that's so simple that it doesn't need to be tested because "nothing can go wrong" or "works for me" and I'm not going to buy into that.

**Comment by IIFE** [comment]
I have linked the other PR to illustrate that this kind of change is needed, not to compare implementations as it's 2+ years old.



> neither of the PRs offer any tests nor any in-depth explanation why the __dllexport is needed to be added exactly in the places where it's being suggested (and/or has a comparison of what the protoc plugin does)

This PR states at the start why the dllexport is needed, and follows the same pattern as the protobuf implementation for messages.



> * At the same time, setting up such tests is non trivial amount of work and such tests require maintenance (and we don't have the bandwidth to do maintain them)



> This only stresses that fact that we'd need a good set of tests for this

The problem with these two statements is that on on hand it's saying you won't accept tests for the plugin because you don't want the maintenance overhead, and on another hand tests are required for this kind of PR.

I'm not against adding tests (I'd rather have tests), and as I said at the start I'm happy to add tests if I was pointed to how the gRPC plugin is being tested. Clearly, there is no test support for the plugin in the first place, which is a repo limitation rather than a problem with the PR. Additionally, previous PRs for the gRPC plugin were accepted without tests.

The whole thing boils down to this - The protobuf plugin allows messages to be generated with a dllexport, the gRPC plugin doesn't, and this essentially makes the gRPC plugin incompatible with the protobuf one. This is simply a bug in the gRPC plugin, because those two systems need to work together. If a gRPC service consume messages exported from a DLL, the service class needs to have the same option for the two to work together.

Not providing support to build gRPC as a DLL is a different thing to what is being addressed here, so I don't believe the same arguments hold. The former is still a big issue for many users, but that a different story altogether.

From a quick glance 3 users have asked to be able to export gRPC service classes, and the longer the issue goes unaddressed, the more users will raise the same issue. If you're hesitant on accepting community PRs, then I believe an internal look at this issue is worth considering, alongside the lack of testing support for the plugin. Given the large user base of Windows applications that rely on DLLs, I disagree that this is a low priority fix.

Thanks.

**Comment by Maju-Ketchup** [comment]
> From a quick glance 3 users have asked to be able to export gRPC service classes, and the longer the issue goes unaddressed, the more users will raise the same issue.

Make it 4

**Comment by netpoetica** [comment]
I too am on Windows attempting to export services as DLLs and use in a separate gRPC server. This seemed quite intuitive to me, I am surprised that this is not possible. But also please forgive that C++ is not my first language and I may not have the depth and breadth of understanding of DLL issues that gRPC may be contending with.

Very simply, my use case is like so:

1) Create a project, "grpc-plugin-example" that builds the ServiceImplementation and a dllexport factory function that instantiates that service implementation into a DLL
2) In my grpc-server project, LoadLibrary("grpc-plugin-example.dll"), get Proc Address of that factory function, instantiate the service, and call builder.RegisterService(&pGrpcPluginExampleService)

At this point, I would expect the service to be running in my gRPC server successfully, but instead, I am running into the following error when making a request from a C++ client to my gRPC server:

```
Exception thrown: read access violation.
**grpc_core::g_static_metadata_slice_ctx** was nullptr.

Unhandled exception thrown: read access violation.
**grpc_core::g_static_metadata_slice_ctx** was nullptr.
```

The service implementation that is exported looks like:

```
grpc::Status PingPongServiceImpl::PingUnary(
  grpc::ServerContext* context,
  const ping_pong::PingPongRequest* request,
  ping_pong::PingPongReply* reply
) {
  std::cout << "PingPong" << std::endl;

  std::cout
    << "PingPong - input_msg = "
    << request->input_msg()
    << std::endl;

  if (request->input_msg() == "hello") {
    reply->set_output_msg("world");
  } else {
    reply->set_output_msg("I can't pong unless you ping me 'hello'!");
  }

  std::cout << "Replying with " << reply->output_msg() << std::endl;

  return grpc::Status::OK;
}

extern "C" __declspec(dllexport) void __cdecl CreateNapaService(grpc::Service** service) {
  *service = new PingPongServiceImpl();
```

It is difficult for me to understand how this would not be a common use case?

**Comment by malkia** [comment]
My recent use case was to expose OpenTelemetry protos into relatively small library for our C++ users.

For comparison, if compiled with statically with vcpkg for Debug/Release x64 the static libs are 3.4GB (grpc, protobuf, upb, zlib, c-ares, openssl, etc.).

Another issue is that by statically linking we may run into symbol clashes, or worst silent ODR violations. The rest of our tools pipeline also uses some of the libraries, but at different versions.

So I've tasked myself with creating small "sdk" to fill the needs.

Compiled the generated opentelemtry *.proto files with dllexport_decl=OPENTEL_API
From there I've got a 8mb .dll file with the protos I need, including various Request / Response message.
But then (as the rest) realize I can't do that for gRPC.

I ended up with this "solution", and in my case it's doable:
I've compiled one .dll (OpenTel.dll) with the protos+grpc+protobuf (static versions of the libs, only exporting the opentelemetry-protos).
Then I've had another libproto.dll (+ zlib.dll) at around 2mb.
Excluding the .pdbs this gives a small "sdk" to C++ clients where they can create metrics, and form the request/response entries.
As for the actual sends, since I'm using the one-off calls, I've exposed multiple small shim versions as exported "C++" functions in that OpenTel.dll.

SO basically, the client instead of directly calling and initiaiting gRPC on it's side, it has to call the shim.
Still though the nasty issue with protos being filled on the client side (and not yet another "shim") was avoided.

But ideally, if gRPC allowed a .dll version that would've not be needed.

So how did .NET Framework worked with grpc exposed as C++, what about other languages/runtimes where this was done too?
